core/src/services/onedrive/backend.rs

```diff
@@ -0,0 +1,209 @@
+        Self {
+            root,
+            access_token,
+            client: http_client,
+        }
+    }
+}
+
+impl Debug for OnedriveBackend {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut de = f.debug_struct("OneDriveBackend");
+        de.field("root", &self.root);
+        de.field("access_token", &self.access_token);
+        de.finish()
+    }
+}
+
+#[async_trait]
+impl Accessor for OnedriveBackend {
+    type Reader = IncomingAsyncBody;
+    type BlockingReader = ();
+    type Writer = OneDriveWriter;
+    type BlockingWriter = ();
+    type Pager = ();
+    type BlockingPager = ();
+
+    fn info(&self) -> AccessorInfo {
+        let mut ma = AccessorInfo::default();
+        ma.set_scheme(crate::Scheme::Onedrive)
+            .set_root(&self.root)
+            .set_capability(Capability {
+                read: true,
+                read_can_next: true,
+                write: true,
+                list: true,
+                copy: true,
+                rename: true,
+                ..Default::default()
+            });
+
+        ma
+    }
+
+    async fn read(&self, path: &str, _args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let resp = self.onedrive_get(path).await?;
+
+        let status = resp.status();
+
+        if status.is_redirection() {
+            let headers = resp.headers();
+            let location = parse_location(headers)?;
+
+            match location {
+                None => {
+                    return Err(Error::new(
+                        ErrorKind::ContentIncomplete,
+                        "redirect location not found in response",
+                    ));
+                }
+                Some(location) => {
+                    let resp = self.onedrive_get_redirection(location).await?;
+                    let meta = parse_into_metadata(path, resp.headers())?;
+                    Ok((RpRead::with_metadata(meta), resp.into_body()))
+                }
+            }
+        } else {
+            match status {
+                StatusCode::OK | StatusCode::PARTIAL_CONTENT => {
+                    let meta = parse_into_metadata(path, resp.headers())?;
+                    Ok((RpRead::with_metadata(meta), resp.into_body()))
+                }
+
+                _ => Err(parse_error(resp).await?),
+            }
+        }
+    }
+
+    async fn write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::Writer)> {
+        if args.content_length().is_none() {
+            return Err(Error::new(
+                ErrorKind::Unsupported,
+                "write without content length is not supported",
+            ));
+        }
+
+        let path = build_rooted_abs_path(&self.root, path);
+
+        Ok((
+            RpWrite::default(),
+            OneDriveWriter::new(self.clone(), args, path),
+        ))
+    }
+}
+
+impl OnedriveBackend {
```

Comment:
I perfer write them directly which make the code more clear:

```rust
let url: String = format!(
    "{}{}{}",
    OnedriveBackend::ONEDRIVE_ENDPOINT_PREFIX,
    percent_encode_path(&path),
    OnedriveBackend::ONEDRIVE_ENDPOINT_SUFFIX
);
```

to

```rust
let url: String = format!(
    "https://graph.microsoft.com/v1.0/me/drive/root:{}:/content",
    percent_encode_path(&path),
);
```

Reply:
updated